modules/amm-executor/src/main/scala/org/ergoplatform/dex/executor/amm/services/CFMMBacklog.scala

```diff
@@ -0,0 +1,202 @@
+    _                              <- isoKGI.to(recoverPendingQueue[F](pendingQueueRef, cfmmOrders))
+  } yield BacklogConfig.access
+    .map(cfg =>
+      new CFMMBacklogTracingMid[F] attach new Live[F](
+        pendingQueueRef,
+        suspendedQueueRef,
+        revisitOrders,
+        cfmmOrders,
+        cfg
+      )
+    )
+    .embed
+
+  final private class Live[F[_]: Monad: Clock: GenRandom](
+    pendingQueueRef: Atom[F, mutable.PriorityQueue[WeightedOrder]],
+    suspendedQueueRef: Atom[F, mutable.PriorityQueue[WeightedOrder]],
+    revisitOrders: Atom[F, List[WeightedOrder]],
+    cfmmOrders: CFMMOrders[F],
+    backlogConfig: BacklogConfig
+  ) extends CFMMBacklog[F] {
+
+    /** Put an order to the backlog.
+      */
+    def put(order: CFMMOrder): F[Unit] =
+      cfmmOrders.put(order) >> pendingQueueRef.update(_ += WeightedOrder.fromOrder(order))
+
+    /** Put an order with priceTooHigh or priceTooLow execution err
+      */
+    def suspend(order: CFMMOrder): F[Unit] =
+      suspendedQueueRef.update(_ += WeightedOrder.fromOrder(order))
+
+    /** Put possibly executed order to the backlog
+      */
+    def checkLater(order: CFMMOrder): F[Unit] =
+      revisitOrders.update(_ :+ WeightedOrder.fromOrder(order))
+
+    /** Get candidate order for execution. Blocks until an order is available.
+      */
+    def get: F[Option[CFMMOrder]] = for {
+      _      <- filterRevisit
+      random <- GenRandom.nextInt[F](100)
+      order <-
+        if (random > backlogConfig.suspendedProbability)
+          getMaxOrderFromQueue(pendingQueueRef)
+        else
+          getMaxOrderFromQueue(suspendedQueueRef)
+    } yield order
+
+    /** Put an order from the backlog.
+      */
+    def drop(id: OrderId): F[Unit] =
+      cfmmOrders.drop(id)
+
+    def getMaxOrderFromQueue(queue: Atom[F, mutable.PriorityQueue[WeightedOrder]]): F[Option[CFMMOrder]] =
+      for {
+        maxId <- queue.modify { queue =>
+                   val elem = if (queue.isEmpty) none else queue.dequeue().some
+                   (queue, elem)
+                 }
+        time <- now.millis
+        suspendedElem <- maxId match {
+                           case Some(value) if (time - value.timestamp) < backlogConfig.orderLifetime.toMillis =>
+                             cfmmOrders.get(value.orderId) >>= {
+                               case Some(value) => value.some.pure[F]
+                               case None        => getMaxOrderFromQueue(queue)
+                             }
+                           case Some(value) =>
+                             cfmmOrders.drop(value.orderId) >> getMaxOrderFromQueue(queue)
+                           case None => none.pure[F]
+                         }
+      } yield suspendedElem
+
+    def filterRevisit: F[Unit] = for {
+      curTime <- now.millis
+      possible2pending <-
+        revisitOrders.modify(_.span(wOrd => (curTime - wOrd.timestamp > backlogConfig.orderExecutionTime.toMillis)))
```

Comment:
I am a little bit confused here. `span` returns 2 lists - first contains everything until predicate p (`curTime - wOrd.timestamp > backlogConfig.orderExecutionTime.toMillis`), there one is everything after this predicate. And `modify` saves the first one list (with bigger timestamp than config) and returns the second one (everything that has less timestamp that config). But I think we should return in our case the list with all orders with timestamp bigger than config value, is not it?

Reply:
yes, incorrect condition